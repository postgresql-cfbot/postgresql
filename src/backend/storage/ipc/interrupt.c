/*-------------------------------------------------------------------------
 *
 * interrupt.c
 *	  Interrupt handling routines.
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *	  src/backend/storage/interrupt.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <unistd.h>

#include "miscadmin.h"
#include "port/atomics.h"
#include "storage/interrupt.h"
#include "storage/ipc.h"
#include "storage/proc.h"
#include "storage/procsignal.h"
#include "storage/waiteventset.h"
#include "utils/guc.h"
#include "utils/memutils.h"

static pg_atomic_uint32 LocalPendingInterrupts;
static pg_atomic_uint32 LocalMaybeSleepingOnInterrupts;

pg_atomic_uint32 *MyPendingInterrupts = &LocalPendingInterrupts;
pg_atomic_uint32 *MyMaybeSleepingOnInterrupts = &LocalMaybeSleepingOnInterrupts;

/*
 * Switch to local interrupts.  Other backends can't send interrupts to this
 * one.  Only RaiseInterrupt() can set them, from inside this process.
 */
void
SwitchToLocalInterrupts(void)
{
	if (MyPendingInterrupts == &LocalPendingInterrupts)
		return;

	MyPendingInterrupts = &LocalPendingInterrupts;
	MyMaybeSleepingOnInterrupts = &LocalMaybeSleepingOnInterrupts;

	/*
	 * Make sure that SIGALRM handlers that call RaiseInterrupt() are now
	 * seeing the new MyPendingInterrupts destination.
	 */
	pg_memory_barrier();

	/*
	 * Mix in the interrupts that we have received already in our shared
	 * interrupt vector, while atomically clearing it.  Other backends may
	 * continue to set bits in it after this point, but we've atomically
	 * transferred the existing bits to our local vector so we won't get
	 * duplicated interrupts later if we switch backx.
	 */
	pg_atomic_fetch_or_u32(MyPendingInterrupts,
						   pg_atomic_exchange_u32(&MyProc->pendingInterrupts, 0));
}

/*
 * Switch to shared memory interrupts.  Other backends can send interrupts to
 * this one if they know its ProcNumber, and we'll now see any that we missed.
 */
void
SwitchToSharedInterrupts(void)
{
	if (MyPendingInterrupts == &MyProc->pendingInterrupts)
		return;

	MyPendingInterrupts = &MyProc->pendingInterrupts;
	MyMaybeSleepingOnInterrupts = &MyProc->maybeSleepingOnInterrupts;

	/*
	 * Make sure that SIGALRM handlers that call RaiseInterrupt() are now
	 * seeing the new MyPendingInterrupts destination.
	 */
	pg_memory_barrier();

	/* Mix in any unhandled bits from LocalPendingInterrupts. */
	pg_atomic_fetch_or_u32(MyPendingInterrupts,
						   pg_atomic_exchange_u32(&LocalPendingInterrupts, 0));
}

/*
 * Set an interrupt flag in this backend.
 */
void
RaiseInterrupt(InterruptType reason)
{
	pg_atomic_fetch_or_u32(MyPendingInterrupts, 1 << reason);
	WakeupMyProc();
}

/*
 * Set an interrupt flag in another backend.
 */
void
SendInterrupt(InterruptType reason, ProcNumber pgprocno)
{
	PGPROC	   *proc;

	Assert(pgprocno != INVALID_PROC_NUMBER);
	Assert(pgprocno >= 0);
	Assert(pgprocno < ProcGlobal->allProcCount);

	proc = &ProcGlobal->allProcs[pgprocno];
	pg_atomic_fetch_or_u32(&proc->pendingInterrupts, 1 << reason);
	WakeupOtherProc(proc);
}
