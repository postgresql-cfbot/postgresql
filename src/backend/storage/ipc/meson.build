backend_sources += files(
  'barrier.c',
  'dsm.c',
  'dsm_impl.c',
  'ipc.c',
  'ipci.c',
  'latch.c',
  'pmsignal.c',
  'procarray.c',
  'procsignal.c',
  'shm_mq.c',
  'shm_toc.c',
  'shmem.c',
  'shmqueue.c',
  'signalfuncs.c',
  'sinval.c',
  'sinvaladt.c',
  'standby.c',

)
