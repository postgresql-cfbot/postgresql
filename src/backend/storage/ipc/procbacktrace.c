/*-------------------------------------------------------------------------
 *
 * procbacktrace.c
 *	  Backtrace-related routines
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *	  src/backend/storage/ipc/procbacktrace.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#ifdef HAVE_EXECINFO_H
#include <execinfo.h>
#endif

#include "funcapi.h"
#include "miscadmin.h"
#include "storage/proc.h"
#include "storage/procarray.h"
#include "storage/procbacktrace.h"
#include "storage/procsignal.h"
#include "utils/fmgrprotos.h"

#ifdef HAVE_EXECINFO_H
static volatile sig_atomic_t backtrace_functions_loaded = false;
#endif

/*
 * Handle receipt of an interrupt indicating logging of current backtrace.
 *
 * We capture the backtrace within this signal handler itself in a safe manner.
 * It is ensured that no memory allocations happen here. This function emits
 * backtrace to stderr as writing to server log allocates memory, making the
 * signal halder unsafe. Also, the shared library implementing the
 * backtrace-related functions is preloaded to avoid memory allocations upon
 * first-time loading, see LoadBacktraceFunctions.
 */
void
HandleLogBacktraceInterrupt(void)
{
#ifdef HAVE_EXECINFO_H
	void	   *buf[100];
	int			nframes;

	if (!backtrace_functions_loaded)
		return;

	nframes = backtrace(buf, lengthof(buf));

	write_stderr("logging current backtrace of process with PID %d:\n",
				 MyProcPid);
	backtrace_symbols_fd(buf, nframes, fileno(stderr));
#endif
}

/*
 * Load backtrace shared library.
 *
 * Any backtrace-related functions when called for the first time dynamically
 * loads the shared library, which usually triggers a call to malloc, making
 * them unsafe to use in signal handlers. This function makes backtrace-related
 * functions signal-safe.
 *
 * Note that this function is supposed to be called in the early life of a
 * process, preferably after signal handlers are setup; but not from within a
 * signal handler.
 */
void
LoadBacktraceFunctions(void)
{
#ifdef HAVE_BACKTRACE_SYMBOLS
	void	   *buf[2];

	if (backtrace_functions_loaded)
		return;

	/*
	 * It is enough to call any one backtrace-related function to ensure that
	 * the corresponding shared library is dynamically loaded. We just load
	 * two most recent function calls, as we don't use the backtrace anyway.
	 */
	backtrace(buf, lengthof(buf));

	backtrace_functions_loaded = true;
#endif
}

/*
 * pg_log_backend_backtrace
 *		Signal a backend or an auxiliary process to log its current backtrace.
 *
 * By default, only superusers are allowed to signal to log the backtrace
 * because allowing any users to issue this request at an unbounded
 * rate would cause lots of log messages on stderr and which can lead to
 * denial of service. Additional roles can be permitted with GRANT.
 *
 * On receipt of this signal, a backend or an auxiliary process emits the
 * current backtrace to stderr in the signal handler.
 */
Datum
pg_log_backend_backtrace(PG_FUNCTION_ARGS)
{
#ifdef HAVE_BACKTRACE_SYMBOLS
	int			pid = PG_GETARG_INT32(0);
	PGPROC	   *proc;
	BackendId	backendId = InvalidBackendId;

	proc = BackendPidGetProc(pid);

	/*
	 * See if the process with given pid is a backend or an auxiliary process.
	 *
	 * If the given process is a backend, use its backend id in
	 * SendProcSignal() later to speed up the operation. Otherwise, don't do
	 * that because auxiliary processes (except the startup process) don't
	 * have a valid backend id.
	 */
	if (proc != NULL)
		backendId = proc->backendId;
	else
		proc = AuxiliaryPidGetProc(pid);

	/*
	 * BackendPidGetProc() and AuxiliaryPidGetProc() return NULL if the pid
	 * isn't valid; but by the time we reach kill(), a process for which we
	 * get a valid proc here might have terminated on its own.  There's no way
	 * to acquire a lock on an arbitrary process to prevent that. But since
	 * this mechanism is usually used to debug a backend or an auxiliary
	 * process running and consuming lots of memory, that it might end on its
	 * own first and its memory contexts are not logged is not a problem.
	 */
	if (proc == NULL)
	{
		/*
		 * This is just a warning so a loop-through-resultset will not abort
		 * if one backend terminated on its own during the run.
		 */
		ereport(WARNING,
				(errmsg("PID %d is not a PostgreSQL server process", pid)));
		PG_RETURN_BOOL(false);
	}

	if (SendProcSignal(pid, PROCSIG_LOG_BACKTRACE, backendId) < 0)
	{
		/* Again, just a warning to allow loops */
		ereport(WARNING,
				(errmsg("could not send signal to process %d: %m", pid)));
		PG_RETURN_BOOL(false);
	}

	PG_RETURN_BOOL(true);
#else
	ereport(WARNING,
			errmsg("backtrace generation is not supported by this installation"),
			errhint("You need to build PostgreSQL with library containing backtrace_symbols."));

	PG_RETURN_BOOL(false);
#endif
}
