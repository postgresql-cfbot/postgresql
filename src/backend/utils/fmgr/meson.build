backend_sources += files(
  'dfmgr.c',
  'fmgr.c',
  'funcapi.c',
)

# fmgrtab.c
generated_backend_sources += fmgrtab_target[2]
