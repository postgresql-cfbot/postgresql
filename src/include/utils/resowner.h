/*-------------------------------------------------------------------------
 *
 * resowner.h
 *	  POSTGRES resource owner definitions.
 *
 * Query-lifespan resources are tracked by associating them with
 * ResourceOwner objects.  This provides a simple mechanism for ensuring
 * that such resources are freed at the right time.
 * See utils/resowner/README for more info.
 *
 *
 * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/utils/resowner.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef RESOWNER_H
#define RESOWNER_H


/*
 * ResourceOwner objects are an opaque data structure known only within
 * resowner.c.
 */
typedef struct ResourceOwnerData *ResourceOwner;


/*
 * Globally known ResourceOwners
 */
extern PGDLLIMPORT ResourceOwner CurrentResourceOwner;
extern PGDLLIMPORT ResourceOwner CurTransactionResourceOwner;
extern PGDLLIMPORT ResourceOwner TopTransactionResourceOwner;
extern PGDLLIMPORT ResourceOwner AuxProcessResourceOwner;

/*
 * Resource releasing is done in three phases: pre-locks, locks, and
 * post-locks.  The pre-lock phase must release any resources that are
 * visible to other backends (such as pinned buffers); this ensures that
 * when we release a lock that another backend may be waiting on, it will
 * see us as being fully out of our transaction.  The post-lock phase
 * should be used for backend-internal cleanup.
 */
typedef enum
{
	RESOURCE_RELEASE_BEFORE_LOCKS,
	RESOURCE_RELEASE_LOCKS,
	RESOURCE_RELEASE_AFTER_LOCKS
} ResourceReleasePhase;

/*
 * In order to track an object, resowner.c needs a few callbacks for it.
 * The callbacks for an resource of a specific kind are encapsulated in
 * ResourceOwnerFuncs.
 *
 * Note that the callback occurs post-commit or post-abort, so these callback
 * functions can only do noncritical cleanup.
 */
typedef struct ResourceOwnerFuncs
{
	const char *name;		/* name for the object kind, for debugging */

	ResourceReleasePhase phase; /* when are these objects released? */

	/*
	 * Release resource.
	 *
	 * NOTE: this must call ResourceOwnerForget to disassociate it with the
	 * resource owner.
	 */
	void (*ReleaseResource)(Datum res);

	/*
	 * Print a warning, when a resource has not been properly released before
	 * commit.
	 */
	void (*PrintLeakWarning)(Datum res);

} ResourceOwnerFuncs;

/*
 *	Dynamically loaded modules can get control during ResourceOwnerRelease
 *	by providing a callback of this form.
 */
typedef void (*ResourceReleaseCallback) (ResourceReleasePhase phase,
										 bool isCommit,
										 bool isTopLevel,
										 void *arg);


/*
 * Functions in resowner.c
 */

/* generic routines */
extern ResourceOwner ResourceOwnerCreate(ResourceOwner parent,
										 const char *name);
extern void ResourceOwnerRelease(ResourceOwner owner,
								 ResourceReleasePhase phase,
								 bool isCommit,
								 bool isTopLevel);
extern void ResourceOwnerDelete(ResourceOwner owner);
extern ResourceOwner ResourceOwnerGetParent(ResourceOwner owner);
extern void ResourceOwnerNewParent(ResourceOwner owner,
								   ResourceOwner newparent);

extern void ResourceOwnerEnlarge(ResourceOwner owner);
extern void ResourceOwnerRemember(ResourceOwner owner, Datum res, ResourceOwnerFuncs *kind);
extern void ResourceOwnerForget(ResourceOwner owner, Datum res, ResourceOwnerFuncs *kind);

extern void RegisterResourceReleaseCallback(ResourceReleaseCallback callback,
											void *arg);
extern void UnregisterResourceReleaseCallback(ResourceReleaseCallback callback,
											  void *arg);

extern void CreateAuxProcessResourceOwner(void);
extern void ReleaseAuxProcessResources(bool isCommit);

/* special support for local lock management */
struct LOCALLOCK;
extern void ResourceOwnerRememberLock(ResourceOwner owner, struct LOCALLOCK *locallock);
extern void ResourceOwnerForgetLock(ResourceOwner owner, struct LOCALLOCK *locallock);

/* special function to relase all plancache references */
extern void ResourceOwnerReleaseAllPlanCacheRefs(ResourceOwner owner);

#endif							/* RESOWNER_H */
