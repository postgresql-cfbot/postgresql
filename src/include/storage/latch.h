/*-------------------------------------------------------------------------
 *
 * latch.h
 *	  Backwards-compatibility macros for the old Latch interface
 *
 * Latches were an inter-process signalling mechanism that was replaced
 * in PostgreSQL verson 18 with Interrupts, see "storage/interrupt.h".
 * This file contains macros that map old Latch calls to the new interface.
 * The mapping is not perfect, it only covers the basic usage of setting
 * and waiting for the current process's own latch (MyLatch), which is
 * mapped to raising or waiting for INTERRUPT_GENERAL_WAKUP.
 *
 * The WaitEventSet functions that used to be here are now in
 * "storage/waitevents.h".
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/storage/latch.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef LATCH_H
#define LATCH_H

#include "storage/interrupt.h"

#define WL_LATCH_SET WL_INTERRUPT

/*
 * These compatibility macros only support operating on MyLatch. It used to
 * be a pointer to a Latch struct, but now it's just a dummy int variable.
 */

#define MyLatch ((void *) 0xAAAA)

static inline void
SetLatch(void *dummy)
{
	Assert(dummy == MyLatch);
	RaiseInterrupt(INTERRUPT_GENERAL_WAKEUP);
}

static inline void
ResetLatch(void *dummy)
{
	Assert(dummy == MyLatch);
	ClearInterrupt(INTERRUPT_GENERAL_WAKEUP);
}

static inline int
WaitLatch(void *dummy, int wakeEvents, long timeout, uint32 wait_event_info)
{
	uint32		interrupt_mask = 0;

	Assert(dummy == MyLatch || dummy == NULL);

	if ((wakeEvents & WL_LATCH_SET) && dummy == MyLatch)
		interrupt_mask = 1 << INTERRUPT_GENERAL_WAKEUP;
	return WaitInterrupt(interrupt_mask, wakeEvents,
						 timeout, wait_event_info);
}

static inline int
WaitLatchOrSocket(void *dummy, int wakeEvents, pgsocket sock, long timeout,
				  uint32 wait_event_info)
{
	uint32		interrupt_mask = 0;

	Assert(dummy == MyLatch || dummy == NULL);

	if ((wakeEvents & WL_LATCH_SET) && dummy == MyLatch)
		interrupt_mask = 1 << INTERRUPT_GENERAL_WAKEUP;
	return WaitInterruptOrSocket(interrupt_mask, wakeEvents,
								 sock, timeout, wait_event_info);
}

#endif
