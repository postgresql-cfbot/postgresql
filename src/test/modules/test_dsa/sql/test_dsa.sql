CREATE EXTENSION test_dsa;

SELECT test_dsa_random(3, 5, 16, 10240, 'random');
SELECT test_dsa_random(3, 5, 16, 10240, 'forwards');
SELECT test_dsa_random(3, 5, 16, 10240, 'backwards');

SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'random', 5);
SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'forwards', 5);
SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'backwards', 5);

SELECT test_dsa_oom(1024);
SELECT test_dsa_oom(8192);
SELECT test_dsa_oom(10240);