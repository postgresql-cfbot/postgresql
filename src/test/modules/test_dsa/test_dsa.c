/* -------------------------------------------------------------------------
 *
 * test_dsa.c
 *		Simple exercises for dsa.c.
 *
 * Copyright (C) 2016-2023, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		src/test/modules/test_dsa/test_dsa.c
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"

#include "fmgr.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "postmaster/bgworker.h"
#include "storage/ipc.h"
#include "storage/latch.h"
#include "storage/proc.h"
#include "utils/builtins.h"
#include "utils/dsa.h"
#include "utils/resowner.h"
#include "utils/timestamp.h"

#include <stdlib.h>
#include <unistd.h>

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(test_dsa_random);
PG_FUNCTION_INFO_V1(test_dsa_random_parallel);
PG_FUNCTION_INFO_V1(test_dsa_oom);

PGDLLEXPORT void test_dsa_random_worker_main(Datum arg);

/* Which order to free objects in, within each loop. */
typedef enum
{
	/* Free in random order. */
	MODE_RANDOM,
	/* Free in the same order we allocated (FIFO). */
	MODE_FORWARDS,
	/* Free in reverse order of allocation (LIFO). */
	MODE_BACKWARDS
} test_mode;

/* Per-worker results. */
typedef struct
{
	pid_t		pid;
	int64		count;
	TimeOffset	elapsed_time;
} test_result;

/* Parameters for a test run, passed to workers. */
typedef struct
{
	int			loops;
	int			num_allocs;
	int			min_alloc;
	int			max_alloc;
	test_mode	mode;
	test_result results[FLEXIBLE_ARRAY_MEMBER];
} test_parameters;

/* The startup message given to each worker. */
typedef struct
{
	/* How to connect to the shmem area. */
	dsa_handle	area_handle;
	/* Where to find the parameters. */
	dsa_pointer parameters;
	/* What index this worker should write results to. */
	size_t		output_index;
} test_hello;

static test_mode
parse_test_mode(text *mode)
{
	test_mode	result = MODE_RANDOM;
	char	   *cstr = text_to_cstring(mode);

	if (strcmp(cstr, "random") == 0)
		result = MODE_RANDOM;
	else if (strcmp(cstr, "forwards") == 0)
		result = MODE_FORWARDS;
	else if (strcmp(cstr, "backwards") == 0)
		result = MODE_BACKWARDS;
	else
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("unknown mode")));
	return result;
}

static void
check_parameters(const test_parameters *parameters)
{
	if (parameters->min_alloc < 1 || parameters->min_alloc > parameters->max_alloc)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("min_alloc must be >= 1, and min_alloc must be <= max_alloc")));
}

static int
my_tranche_id(void)
{
	static int	tranche_id = 0;

	if (tranche_id == 0)
		tranche_id = LWLockNewTrancheId();

	return tranche_id;
}

static void
do_random_test(dsa_area *area, size_t output_index, test_parameters *parameters)
{
	dsa_pointer *objects;
	int			min_alloc;
	int			extra_alloc;
	int32		i;
	int32		loop;
	TimestampTz start_time = GetCurrentTimestamp();
	int64		total_allocations = 0;

	/*
	 * Make tests reproducible (on the same computer at least) by using the
	 * same random sequence every time.
	 */
	srand(42);

	min_alloc = parameters->min_alloc;
	extra_alloc = parameters->max_alloc - parameters->min_alloc;

	objects = palloc(sizeof(dsa_pointer) * parameters->num_allocs);
	Assert(objects != NULL);
	for (loop = 0; loop < parameters->loops; ++loop)
	{
		int			num_actually_allocated = 0;

		for (i = 0; i < parameters->num_allocs; ++i)
		{
			size_t		size;
			void	   *memory;

			/* Adjust size randomly if needed. */
			size = min_alloc;
			if (extra_alloc > 0)
				size += rand() % extra_alloc;

			/* Allocate! */
			objects[i] = dsa_allocate_extended(area, size, DSA_ALLOC_NO_OOM);
			if (!DsaPointerIsValid(objects[i]))
			{
				elog(LOG, "dsa: loop %d: out of memory after allocating %d objects", loop, i + 1);
				break;
			}
			++num_actually_allocated;
			/* Pay the cost of accessing that memory */
			memory = dsa_get_address(area, objects[i]);
			memset(memory, 42, size);
		}
		if (parameters->mode == MODE_RANDOM)
		{
			for (i = 0; i < num_actually_allocated; ++i)
			{
				size_t		x = rand() % num_actually_allocated;
				size_t		y = rand() % num_actually_allocated;
				dsa_pointer temp = objects[x];

				objects[x] = objects[y];
				objects[y] = temp;
			}
		}
		if (parameters->mode == MODE_BACKWARDS)
		{
			for (i = num_actually_allocated - 1; i >= 0; --i)
				dsa_free(area, objects[i]);
		}
		else
		{
			for (i = 0; i < num_actually_allocated; ++i)
				dsa_free(area, objects[i]);
		}
		total_allocations += num_actually_allocated;
	}
	pfree(objects);

	parameters->results[output_index].elapsed_time = GetCurrentTimestamp() - start_time;
	parameters->results[output_index].pid = MyProcPid;
	parameters->results[output_index].count = total_allocations;
}

/* Non-parallel version: just do it. */
Datum
test_dsa_random(PG_FUNCTION_ARGS)
{
	test_parameters *parameters;
	dsa_area   *area;

	parameters =
		palloc(offsetof(test_parameters, results) + sizeof(test_result));
	parameters->loops = PG_GETARG_INT32(0);
	parameters->num_allocs = PG_GETARG_INT32(1);
	parameters->min_alloc = PG_GETARG_INT32(2);
	parameters->max_alloc = PG_GETARG_INT32(3);
	parameters->mode = parse_test_mode(PG_GETARG_TEXT_PP(4));
	check_parameters(parameters);

	area = dsa_create(my_tranche_id());
	do_random_test(area, 0, parameters);
	dsa_dump(area);
	dsa_detach(area);

	pfree(parameters);

	PG_RETURN_NULL();
}

void
test_dsa_random_worker_main(Datum arg)
{
	test_hello	hello;
	dsa_area   *area;
	test_parameters *parameters;

	CurrentResourceOwner = ResourceOwnerCreate(NULL, "test_dsa toplevel");

	/* Receive hello message and attach to shmem area. */
	memcpy(&hello, MyBgworkerEntry->bgw_extra, sizeof(hello));
	area = dsa_attach(hello.area_handle);
	Assert(area != NULL);
	parameters = dsa_get_address(area, hello.parameters);
	Assert(parameters != NULL);

	do_random_test(area, hello.output_index, parameters);

	dsa_detach(area);
}

/* Parallel version: fork a bunch of background workers to do it. */
Datum
test_dsa_random_parallel(PG_FUNCTION_ARGS)
{
	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
	TupleDesc	tupdesc;
	Tuplestorestate *tupstore;
	MemoryContext per_query_ctx;
	MemoryContext oldcontext;

	test_hello	hello;
	test_parameters *parameters;
	dsa_area   *area;
	int			workers;
	int			i;
	BackgroundWorkerHandle **handles;

	/* tuplestore boilerplate stuff... */
	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("set-valued function called in context that cannot accept a set")));
	if (!(rsinfo->allowedModes & SFRM_Materialize))
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("materialize mode required, but it is not " \
						"allowed in this context")));
	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
		elog(ERROR, "return type must be a row type");
	per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
	oldcontext = MemoryContextSwitchTo(per_query_ctx);
	tupstore = tuplestore_begin_heap(true, false, work_mem);
	rsinfo->returnMode = SFRM_Materialize;
	rsinfo->setResult = tupstore;
	rsinfo->setDesc = tupdesc;
	MemoryContextSwitchTo(oldcontext);

	/* Prepare to work! */
	workers = PG_GETARG_INT32(5);
	handles = palloc(sizeof(BackgroundWorkerHandle *) * workers);

	/* Set up the shared memory area. */
	area = dsa_create(my_tranche_id());

	/* The workers then will attach to it. */
	hello.area_handle = dsa_get_handle(area);

	/* Allocate space for the parameters object. */
	hello.parameters = dsa_allocate(area,
									offsetof(test_parameters, results) +
									sizeof(test_result) * workers);
	Assert(DsaPointerIsValid(hello.parameters));

	/* Set up the parameters object. */
	parameters = dsa_get_address(area, hello.parameters);
	parameters->loops = PG_GETARG_INT32(0);
	parameters->num_allocs = PG_GETARG_INT32(1);
	parameters->min_alloc = PG_GETARG_INT32(2);
	parameters->max_alloc = PG_GETARG_INT32(3);
	parameters->mode = parse_test_mode(PG_GETARG_TEXT_PP(4));
	check_parameters(parameters);

	/* Start the workers. */
	for (i = 0; i < workers; ++i)
	{
		BackgroundWorker bgw;

		memset(&bgw, 0, sizeof(bgw));
		snprintf(bgw.bgw_name, sizeof(bgw.bgw_name), "worker%d", i);
		bgw.bgw_flags = BGWORKER_SHMEM_ACCESS;
		bgw.bgw_start_time = BgWorkerStart_PostmasterStart;
		bgw.bgw_restart_time = BGW_NEVER_RESTART;
		snprintf(bgw.bgw_library_name, sizeof(bgw.bgw_library_name),
				 "test_dsa");
		snprintf(bgw.bgw_function_name, sizeof(bgw.bgw_function_name),
				 "test_dsa_random_worker_main");
		Assert(sizeof(parameters) <= BGW_EXTRALEN);
		/* Each worker will write its output to a different slot. */
		hello.output_index = i;
		memcpy(bgw.bgw_extra, &hello, sizeof(hello));
		bgw.bgw_notify_pid = MyProcPid;

		if (!RegisterDynamicBackgroundWorker(&bgw, &handles[i]))
			elog(ERROR, "can't start worker");
	}

	/* Wait for the workers to complete. */
	for (i = 0; i < workers; ++i)
	{
		BgwHandleStatus status;

		status = WaitForBackgroundWorkerShutdown(handles[i]);
		if (status == BGWH_POSTMASTER_DIED)
			proc_exit(1);
		Assert(status == BGWH_STOPPED);
	}

	/* Generate result tuples. */
	for (i = 0; i < workers; ++i)
	{
		Datum		values[3];
		bool		nulls[] = {false, false, false};
		Interval   *interval = palloc(sizeof(Interval));

		interval->month = 0;
		interval->day = 0;
		interval->time = parameters->results[i].elapsed_time;

		values[0] = Int32GetDatum(parameters->results[i].pid);
		values[1] = Int64GetDatum(parameters->results[i].count);
		values[2] = PointerGetDatum(interval);
		tuplestore_putvalues(tupstore, tupdesc, values, nulls);
	}
	tuplestore_donestoring(tupstore);

	pfree(handles);
	dsa_detach(area);

	return (Datum) 0;
}

/* Allocate memory until OOM, than free and try allocate again. */
Datum
test_dsa_oom(PG_FUNCTION_ARGS)
{
	test_parameters *parameters;
	dsa_area   *area;
	int64		cnt1,
				cnt2;

	parameters =
		palloc(offsetof(test_parameters, results) + sizeof(test_result));

	parameters->loops = 1;
	parameters->min_alloc = PG_GETARG_INT32(0);
	parameters->max_alloc = parameters->min_alloc;
	check_parameters(parameters);

	parameters->num_allocs = 1024 * 1024 / parameters->min_alloc;
	parameters->mode = MODE_RANDOM;

	/* Cap available memory at 1MB. */
	area = dsa_create(my_tranche_id());
	dsa_set_size_limit(area, 1024 * 1024);
	dsa_dump(area);

	do_random_test(area, 0, parameters);
	dsa_dump(area);
	cnt1 = parameters->results[0].count;

	/* And again... */
	do_random_test(area, 0, parameters);
	dsa_dump(area);
	cnt2 = parameters->results[0].count;

	dsa_detach(area);
	pfree(parameters);

	/* We should have allocated the same amount both times. */
	Assert(cnt1 == cnt2);

	PG_RETURN_NULL();
}
