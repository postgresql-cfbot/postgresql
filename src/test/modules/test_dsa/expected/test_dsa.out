CREATE EXTENSION test_dsa;
SELECT test_dsa_random(3, 5, 16, 10240, 'random');
 test_dsa_random 
-----------------
 
(1 row)

SELECT test_dsa_random(3, 5, 16, 10240, 'forwards');
 test_dsa_random 
-----------------
 
(1 row)

SELECT test_dsa_random(3, 5, 16, 10240, 'backwards');
 test_dsa_random 
-----------------
 
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'random', 5);
 count 
-------
     5
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'forwards', 5);
 count 
-------
     5
(1 row)

SELECT count(*) from test_dsa_random_parallel(3, 5, 16, 10240, 'backwards', 5);
 count 
-------
     5
(1 row)

SELECT test_dsa_oom(1024);
 test_dsa_oom 
--------------
 
(1 row)

SELECT test_dsa_oom(8192);
 test_dsa_oom 
--------------
 
(1 row)

SELECT test_dsa_oom(10240);
 test_dsa_oom 
--------------
 
(1 row)

